### PR TITLE
Enable Multiple Data Sources

### PR DESCRIPTION
**Pull request by SeanLeRoy**
**Description**
This changeset enables users to select more than one data source in a single query. This is largely a proof-of-concept implementation of this feature as I think we could spend more time to make this much more performant like by perhaps creating views that are representations of a query's data sources.

**How to review**
Most of the changes to look for are in the SQLBuilder & logics.tsx

**Testing**
Tested on desktop and in the web adding various types of data source combination and performing queries.

### CODE REVIEW

**Review by aswallace — COMMENTED** [review]
Having some trouble testing locally. Main thing is that when I use two different source files that should have the same columns, it shows up like this with null columns: 
image
The data from these two sources came from the same file, I just downloaded them as two separate files and then re-selected them each as datasources in the same query

Also, if I add a third source I get 
```
Error: Parser Error: syntax error at or near "JOIN"
LINE 5: ...,355,118,721"."File Path") AS "File Path"
            FROM "file-selection-Tue...
```

File: packages/core/components/QueryPart/QueryDataSource.tsx [hunk 1]
```diff
@@ -1,39 +1,101 @@
...
-                };
-            })}
+            onDelete={
+                selectedDataSources.length > 1
+                    ? (dataSource) => dispatch(selection.actions.removeDataSource(dataSource))
+                    : undefined
+            }
+            onRenderAddMenuList={
+                selectedDataSources[0]?.name === AICS_FMS_DATA_SOURCE_NAME
+                    ? undefined
```
[the remark this answers is not in the document]
Comment (BrianWhitneyAI) [reply packages/core/components/QueryPart/QueryDataSource.tsx:52]: Also seeing this

**Comment by BrianWhitneyAI** [comment]
image
Seeing this when trying to load in various datasets locally.

File: packages/core/services/FileService/DatabaseFileService/index.ts [hunk 2]
```diff
@@ -126,13 +131,17 @@ export default class DatabaseFileService implements FileService {
...
-                    .from(this.dataSourceName as string)
+                    .select(
+                        `COALESCE(${this.dataSourceNames
+                            .map((source) => `"${source}"."File Path"`)
+                            .join(", ")}) AS "File Path"`
+                    )
+                    .from(this.dataSourceNames)
                     .whereOr(
                         Object.entries(selection.filters).map(([column, values]) => {
                             const commaSeperatedValues = values.map((v) => `'${v}'`).join(", ");
```
[the remark this answers is not in the document]
Comment (SeanLeRoy) [reply packages/core/services/FileService/DatabaseFileService/index.ts:140]: Fixed 👍

File: packages/core/components/QueryPart/QueryDataSource.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SeanLeRoy) [reply packages/core/components/QueryPart/QueryDataSource.tsx:57]: Fixed 👍

File: packages/web/src/services/DatabaseServiceWeb.ts [hunk 3]
```diff
@@ -33,42 +33,40 @@ export default class DatabaseServiceWeb implements DatabaseService {
...
+                // Default to CSV
+                await connection.query(
+                    `CREATE TABLE "${name}" AS FROM read_csv_auto('${name}', header=true);`
+                );
             }
+            this.existingDataSources.add(name);
+        } finally {
+            await connection.close();
         }
     }
```
[the remark this answers is not in the document]
Comment (SeanLeRoy) [reply packages/web/src/services/DatabaseServiceWeb.ts]: Not seeing this as of this moment, perhaps was fixed incidently?

[the remark this answers is not in the document]
Comment (aswallace) [reply packages/web/src/services/DatabaseServiceWeb.ts]: I'm still having this issue -- when I download multiple files as csv, there is either no header or all the data is in one line in the header

**Comment by aswallace** [comment]
I'm now able to select, filter, group on multiple data sources, both web and desktop. Still having issues with downloading manifests:
Web: 
- I can download successfully for non AICS sources, but when I try to download from AICS source it claims to download successfully, but then the browser gives me a "File wasn't available on site" error and doesn't actually complete the downlaod
Desktop: 
- Still having the header issue with downloading CSVs
- On desktop, I'm still seeing the error Brian mentioned when it tries to reload an external data source, and the modal + error won't go away even if I select a new file

**Comment by SeanLeRoy** [comment]
> I'm now able to select, filter, group on multiple data sources, both web and desktop. Still having issues with downloading manifests: Web:
> 
> * I can download successfully for non AICS sources, but when I try to download from AICS source it claims to download successfully, but then the browser gives me a "File wasn't available on site" error and doesn't actually complete the download
This was an issue with the type I was sending to the databaseservice, should work now.

> 
> Desktop:
> 
> * Still having the header issue with downloading CSVs
Added option "HEADER" to CSV exports, should work

> * I'm still seeing the error Brian mentioned when it tries to reload an external data source, and the modal + error won't go away even if I select a new file
Fixed, the cache of annotations was saving `[]` as a valid list of annotations, seems to be working now

This should all be fixed now.

**Comment by aswallace** [comment]
> This should all be fixed now.

Confirmed that these fixes all worked for me!
Only running into a couple small issues now:
- Desktop: Sometimes when I try to add a new query/data source it adds it to the old/current query rather than creating a new one? I'm not exactly sure how to reproduce
- If the columns aren't all exactly the same for both data sources (e.g., one is missing one column), I get errors (first screenshot is desktop, second is web): 
Screenshot 2024-06-13 at 4 29 09 PM
image


- Saw that when a query with multiple data sources has a source changed, it always throws an error even if it's successful: `Uncaught (in promise) Catalog Error: Table with name AICS-13, AICS-5 does not exist! Did you mean "AICS-13"?  FROM "AICS-13, AICS-5"`

Also two future feature requests:
- would be really nice if we could delete queries without opening them, since clicking on them to delete means the app automatically attempts to run the query, which can be annoying if it's an old data source I don't intend to replace
- Some errors don't render in the file list UI, so it can look like it's just loading forever if you don't know to check console

**Review by aswallace — APPROVED** [review]

**Pull request merged**